patches/tModLoader/Terraria/ModLoader/Config/ConfigManager.cs

```diff
@@ -356,6 +459,138 @@ public static bool EnumerableEquals(IEnumerable a, IEnumerable b)
 		}
 		return !hasNextA && !hasNextB;
 	}
+
+	internal static string FormatTextAttribute(string key, string localization, object[] args)
+	{
+		if (args == null)
+			return localization;
+		for (int i = 0; i < args.Length; i++) {
+			if (args[i] is string s && s.StartsWith("$"))
+				args[i] = Language.GetTextValue(FindKeyInScope(s.Substring(1), key));
+		}
+		return Language.GetText(key).Format(args);
+
+		string FindKeyInScope(string key, string scope)
+		{
+			if (LanguageManager.Instance.Exists(key))
+				return key;
+
+			string[] splitKey = scope.Split(".");
+			for (int j = splitKey.Length - 1; j >= 0; j--) {
+				string partialKey = string.Join(".", splitKey.Take(j + 1));
+				string combinedKey = partialKey + "." + key;
+				if (LanguageManager.Instance.Exists(combinedKey))
+					return combinedKey;
+			}
+
+			return key;
+		}
+	}
+
+	private static T GetAndValidate<T>(MemberInfo memberInfo, bool throwErrors = true, bool errorOnType = false) where T : ConfigKeyAttribute
+	{
+		var configKeyAttribute = (T)Attribute.GetCustomAttribute(memberInfo, typeof(T));
+		if (configKeyAttribute != null) {
+			if (configKeyAttribute.malformed && throwErrors)
+				throw new ValueNotTranslationKeyException($"{nameof(T)} only accepts localization keys for the 'key' parameter.", errorOnType: true);
+		}
+		return configKeyAttribute;
+	}
+
+	// Used to determine which key to register, based only on field/property, not class, not necessarily which key to use in UI.
+	internal static string GetConfigKey<T>(PropertyFieldWrapper memberInfo, ModConfig config, bool fallbackToClass, bool throwErrors, string dataName) where T : ConfigKeyAttribute
+	{
+		var configKeyAttribute = GetAndValidate<T>(memberInfo.MemberInfo, throwErrors: throwErrors, errorOnType: false);
+		if (configKeyAttribute != null) {
+			return configKeyAttribute.key;
+		}
+		configKeyAttribute = GetAndValidate<T>(memberInfo.Type, throwErrors: throwErrors, errorOnType: true);
+		if (configKeyAttribute != null) {
+			if (fallbackToClass) // FinishSetup will catch errors on ConfigKey annotations on classes
+				return configKeyAttribute.key;
+		}
+
+		// Autokey: Determine key from the Type the member belongs to.
+		Type typeMemberBelongsTo = memberInfo.MemberInfo.DeclaringType; 
+		string modName = typeMemberBelongsTo.Assembly.GetName().Name;
+		string className = typeMemberBelongsTo == config.GetType() ? config.Name : typeMemberBelongsTo.Name;
+		if (modName == "tModLoader" || modName == "FNA") // tModLoader keys handle translations for existing classes
+			return $"Config.{className}.{memberInfo.Name}.{dataName}";
+		return $"Mods.{modName}.Configs.{className}.{memberInfo.Name}.{dataName}";
```

Comment:
I notice a lot of complexity coming from needing to thread the `ModConfig` instance through a lot of methods where it is ultimately only used just for `{config.Mod.Name}` or `config.Name` to build paths.

In these code paths, you have to handle both members on a `ModConfig` instance, and members on a custom `Type` used as a config property (recursively).

I notice `ModConfig.Autoload` has a `// TODO: Does non-autoloaded ModConfigs have a use-case?`

The only time when it would be useful for `config.Name` to diverge from `config.GetType().Name` would be if there were two configs with the same type but different names (which is why `Name` is `virtual` in `ModType`)

Now whether we want to ditch manual calls to `Mod.AddConfig` is a different matter, but for localization purposes, I think it's actually _harmful_ to have the generated localization keys depend on `config.Name` rather than just `config.GetType().Name`. The reason being that if you _did_ have two configs with the same type but different names, all the localizations would be duplicated!

If you only have one config instance per type (assume they're singletons), then there is no reason for `config.Name` to not be `config.GetType().Name`

This is all by means of saying, we should do all localization registration on a `Type` basis, and consider `ModConfig` instances by nothing more than their `Type`. This will bring simplicity, consistency and reduce weird behavior in untested cases above.

Also, you can have a `HashSet<Type>` of types you've already registered.

Reply:
implemented.